### PR TITLE
⚠️ Fix object logging in ResourceHasFilterLabel, ResourceNotPaused & ResourceNotPausedAndHasFilterLabel predicates

### LINKED ISSUE #10818: Improve logging of predicates
The predicates we have today only contain an absolute minimum amount of k/v pairs, e.g.
image

Some notes:
* it's impossible to figure out in which controller this predicate is sued
* kind (the 6th value) is just an empty string

I think we have to improve our predicate logs, otherwise they are not really useful.

(I noticed this when trying to debug predicates of a controller, which is very very had if there is no way to determine which log belongs to which controller)

In general we should compare the logs of our predicates with the logs of a regular controller and figure out which additional k/v pairs we need + we should make sure kind is logged correctly

### PR DESCRIPTION
**Pull request by Karthik-K-N**
**What this PR does / why we need it**:

Main motivation of this PR is to log resource kind in predicate logs, In order to accommodate change we need to change the signature of predicates.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Part of #10818

/area utils

### CODE REVIEW

**Comment by Karthik-K-N** [comment]
@sbueringer I have made changes starting with ResourceNotPaused and ResourceHasFilterLabel predicates,  I also ended up changing signature of external.Watch() to take scheme as parameter, Please take a look when you get chance. Thank you.

File: controllers/external/tracker.go [hunk 1]
```diff
@@ -41,7 +42,7 @@ type ObjectTracker struct {
 }
 
 // Watch uses the controller to issue a Watch only if the object hasn't been seen before.
-func (o *ObjectTracker) Watch(log logr.Logger, obj client.Object, handler handler.EventHandler, p ...predicate.Predicate) error {
+func (o *ObjectTracker) Watch(scheme *runtime.Scheme, log logr.Logger, obj client.Object, handler handler.EventHandler, p ...predicate.Predicate) error {
 	// Consider this a no-op if the controller isn't present.
 	if o.Controller == nil {
 		return nil
```
[the remark this answers is not in the document]
Comment (Karthik-K-N) [reply controllers/external/tracker.go]: After adding the nil check for scheme and cache many of the existing test cases are failing, I will fix them in seperate commit.

File: controllers/external/tracker.go [hunk 2]
```diff
@@ -41,7 +42,7 @@ type ObjectTracker struct {
 }
 
 // Watch uses the controller to issue a Watch only if the object hasn't been seen before.
-func (o *ObjectTracker) Watch(log logr.Logger, obj client.Object, handler handler.EventHandler, p ...predicate.Predicate) error {
+func (o *ObjectTracker) Watch(scheme *runtime.Scheme, log logr.Logger, obj client.Object, handler handler.EventHandler, p ...predicate.Predicate) error {
```
[the remark this answers is not in the document]
Comment (Karthik-K-N) [reply controllers/external/tracker.go]: Updated as suggested.

File: controllers/external/tracker.go [hunk 3]
```diff
@@ -47,6 +49,10 @@ func (o *ObjectTracker) Watch(log logr.Logger, obj client.Object, handler handle
 		return nil
 	}
 
+	if o.Cache == nil || o.Scheme == nil {
```
Comment (Karthik-K-N) [inline controllers/external/tracker.go:52]: Added a check here, so it continue to work as no-op when controller is not set and make sure to check these two are set when controller is set.

File: controllers/external/tracker_test.go [hunk 4]
```diff
@@ -64,7 +66,7 @@ func (c *watchCountController) Watch(_ source.Source) error {
 func TestRetryWatch(t *testing.T) {
 	g := NewWithT(t)
 	ctrl := newWatchCountController(true)
-	tracker := ObjectTracker{Controller: ctrl}
+	tracker := ObjectTracker{Controller: ctrl, Scheme: runtime.NewScheme(), Cache: &informertest.FakeInformers{}}
```
Comment (Karthik-K-N) [inline controllers/external/tracker_test.go:69]: Is there any better way to set cache for testing?

**Review by Karthik-K-N — COMMENTED** [review]
Made changes as suggested, have couple of queries please take a look when you get time. Thanks

File: controllers/external/tracker.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (sbueringer) [reply controllers/external/tracker.go]: Yeah but it's pretty horrible that it's a silent no-op

[the remark this answers is not in the document]
Comment (sbueringer) [reply controllers/external/tracker.go]: What do you think about fixing that in a separate PR first?

[the remark this answers is not in the document]
Comment (Karthik-K-N) [reply controllers/external/tracker.go]: Yeah, I can submit a subsequent PR to fix them.

[the remark this answers is not in the document]
Comment (sbueringer) [reply controllers/external/tracker.go]: Okay, sounds fine! Thank you

File: controllers/external/tracker_test.go (hunk 4 above)
Comment (sbueringer) [reply controllers/external/tracker_test.go:69]: I don't think so. The only alternatives I see is to either create a "real" cache or implement our own fake cache.

I would keep the current implementation

**Comment by sbueringer** [comment]
Thank you!

/lgtm

/assign @fabriziopandini @chrischdi

**Review by chrischdi — COMMENTED** [review]
/approve

**Pull request merged**